src/liballoc/vec.rs

```diff
@@ -2000,7 +1999,83 @@ impl<T, I> SpecExtend<T, I> for Vec<T>
         <Vec<T> as SpecExtend<T, I>>::spec_extend(&mut vector, iterator);
         vector
     }
+}
+
+fn from_into_iter_source<T, I>(mut iterator: I) -> Vec<T>
+    where I: Iterator<Item=T> + InPlaceIterable + SourceIter<Source=IntoIter<T>>
+{
+    let mut insert_pos = 0;
+
+    // FIXME: how to drop values written into source when iteration panics?
+    //   tail already gets cleaned by IntoIter::drop
+    while let Some(item) = iterator.next() {
+        let source_iter = iterator.as_inner();
+        let src_buf = source_iter.buf.as_ptr();
+        let src_idx = source_iter.ptr;
+        unsafe {
+            let dst = src_buf.offset(insert_pos as isize);
+            debug_assert!(dst as *const _ < src_idx,
+                          "InPlaceIterable implementation produced more\
+                          items than it consumed from the source");
+            ptr::write(dst, item)
+        }
+        insert_pos += 1;
+    }
+
+    let src = iterator.as_inner();
+    let vec = unsafe { Vec::from_raw_parts(src.buf.as_ptr(), insert_pos, src.cap)};
+    mem::forget(iterator);
```

Comment:
You can't just forget the entire iterator - that will leak the memory of e.g. Map's closure.

Reply:
replaced the forget by overwriting IntoIter fields with the equivalent of an empty vec